core/src/main/scala/akka/persistence/r2dbc/internal/EnvelopeOrigin.scala

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.r2dbc.internal
+
+import akka.annotation.InternalStableApi
+import akka.persistence.query.Offset
+import akka.persistence.query.TimestampOffset
+import akka.persistence.query.UpdatedDurableState
+import akka.persistence.query.typed.EventEnvelope
+
+/**
+ * INTERNAL API
+ */
+@InternalStableApi private[akka] object EnvelopeOrigin {
+  val SourceQuery = ""
+  val SourceBacktracking = "BT"
+  val SourcePubSub = "PS"
+
+  def fromBacktracking(env: EventEnvelope[_]): Boolean =
+    env.source == SourceBacktracking
+
+  def fromBacktracking(change: UpdatedDurableState[_]): Boolean =
+    change.value == null
+
+  def fromPubSub(env: EventEnvelope[_]): Boolean =
+    fromPubSub(env.offset)
```

Comment:
Not `env.source == SourcePubSub` ?

Reply:
right, fixed. nice to get rid of that timestamp check